### ISSUE TITLE
Resizing sometimes causing spurious selections to appear on screen

### ISSUE THREAD

**Issue by ggsalas** [kovidgoyal/kitty#2747]
**Describe the bug**
The screen appears with many cursor blocks that not disappears and makes text not readable.

**To Reproduce**
Can't find a clear way to reproduce the error. Probably happen on resize, fullscreen and "zoom" windows.

**Screenshots**
<img width="1920" alt="Screen Shot 2020-06-07 at 10 24 24" src="https://user-images.githubusercontent.com/8403993/83965970-d0930900-a8b7-11ea-9c7b-c29e23d6d5e8.png">

**Enviroment details**
OS: macos 10.15.4 

```
kitty 0.17.2 created by Kovid Goyal
Darwin gabriels-MacBook-Pro.local 19.4.0 Darwin Kernel Version 19.4.0: Wed Mar  4 22:28:40 PST 2020; root:xnu-6153.101.6~15/RELEASE_X86_64 x86_64
ProductName:	Mac OS X ProductVersion:	10.15.4 BuildVersion:	19E287
Loaded config files: /Users/ggsalas/.config/kitty/kitty.conf
[159 12:12:59.820579] The option macos_hide_titlebar is deprecated. Use hide_window_decorations instead.

Config options different from defaults:
active_border_color     Color(red=98, green=100, blue=131)
active_tab_background   Color(red=233, green=233, blue=244)
active_tab_font_style   (True, False)
active_tab_foreground   Color(red=40, green=41, blue=54)
adjust_line_height      10
allow_remote_control    y
background              Color(red=40, green=41, blue=54)
bold_font               JetBrains Mono Bold
bold_italic_font        JetBrains Mono Bold Italic
color0                  Color(red=40, green=41, blue=54)
color1                  Color(red=234, green=81, blue=178)
color10                 Color(red=58, green=60, blue=78)
color11                 Color(red=77, green=79, blue=104)
color12                 Color(red=98, green=214, blue=232)
color13                 Color(red=241, green=242, blue=248)
color14                 Color(red=0, green=247, blue=105)
color15                 Color(red=233, green=233, blue=244)
color2                  Color(red=235, green=255, blue=135)
color3                  Color(red=0, green=247, blue=105)
color4                  Color(red=98, green=214, blue=232)
color5                  Color(red=180, green=91, blue=207)
color6                  Color(red=161, green=239, blue=228)
color7                  Color(red=233, green=233, blue=244)
color8                  Color(red=98, green=100, blue=131)
color9                  Color(red=180, green=91, blue=207)
cursor                  Color(red=255, green=0, blue=255)
cursor_blink_interval   0.0
cursor_text_color       Color(red=40, green=41, blue=54)
editor                  nvim
enabled_layouts         ['fat:bias=85', 'tall:bias=65', 'stack', 'grid', 'horizontal', 'vertical']
font_family             JetBrains Mono
font_size               14.4
foreground              Color(red=233, green=233, blue=244)
inactive_border_color   Color(red=58, green=60, blue=78)
inactive_tab_background Color(red=58, green=60, blue=78)
inactive_tab_foreground Color(red=98, green=214, blue=232)
inactive_text_alpha     0.7
initial_window_height   (35, 'cells')
initial_window_width    (180, 'cells')
input_delay             6
italic_font             JetBrains Mono Italic
remember_window_size    False
resize_in_steps         True
selection_background    Color(red=233, green=233, blue=244)
selection_foreground    Color(red=40, green=41, blue=54)
sync_to_monitor         False
tab_bar_margin_width    4.0
tab_fade                (1.0, 1.0, 1.0)
url_color               Color(red=98, green=214, blue=232)
window_padding_width    6.0
Added shortcuts:
	control+space KeyAction(func='send_text', args=['all', b'\x10'])
	shift+control+z KeyAction(func='kitten', args=['zoomKitten.py'])
	shift+control+alt+h KeyAction(func='move_window', args=['left'])
	shift+control+alt+j KeyAction(func='move_window', args=['bottom'])
	shift+control+alt+k KeyAction(func='move_window', args=['top'])
	shift+control+alt+l KeyAction(func='move_window', args=['right'])
Changed shortcuts:
	shift+control+0 KeyAction(func='goto_layout', args=['fat:bias=85'])
	shift+control+6 KeyAction(func='goto_layout', args=['vertical'])
	shift+control+7 KeyAction(func='goto_layout', args=['horizontal'])
	shift+control+8 KeyAction(func='goto_layout', args=['grid'])
	shift+control+9 KeyAction(func='goto_layout', args=['tall:bias=65'])
	shift+control+b KeyAction(func='scroll_page_up', args=())
	shift+control+f KeyAction(func='scroll_page_down', args=())
	shift+control+h KeyAction(func='neighboring_window', args=['left'])
	shift+control+j KeyAction(func='neighboring_window', args=['bottom'])
	shift+control+k KeyAction(func='neighboring_window', args=['top'])
	shift+control+l KeyAction(func='neighboring_window', args=['right'])
	shift+control+n KeyAction(func='launch', args=['--env', 'PATH=/usr/local/bin:/Applications/kitty.app/Contents/MacOS:/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Applications/kitty.app/Contents/MacOS:/Users/ggsalas/bin:/Users/ggsalas/go/bin:/Users/ggsalas/.fzf/bin', '--cwd=current', 'nvim', '-c', 'NoteList'])
	shift+control+o KeyAction(func='launch', args=['--env', 'PATH=/usr/local/bin:/Applications/kitty.app/Contents/MacOS:/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Applications/kitty.app/Contents/MacOS:/Users/ggsalas/bin:/Users/ggsalas/go/bin:/Users/ggsalas/.fzf/bin', '--cwd=current', 'nnn'])
	shift+control+t KeyAction(func='new_tab_with_cwd', args=())
	shift+control+left_bracket KeyAction(func='previous_tab', args=())
	shift+control+right_bracket KeyAction(func='next_tab', args=())
	shift+control+enter KeyAction(func='new_window_with_cwd', args=())
	shift+control+right KeyAction(func='move_tab_forward', args=())
	shift+control+left KeyAction(func='move_tab_backward', args=())
	shift+control+down KeyAction(func='resize_window', args=['shorter', 3])
	shift+control+up KeyAction(func='resize_window', args=['taller', 1])
```

**Additional context**
I have started kitty with  --dump-commands and I see this warnings on the console:
<img width="1646" alt="Screen Shot 2020-06-07 at 10 23 59" src="https://user-images.githubusercontent.com/8403993/83966037-2c5d9200-a8b8-11ea-8a30-4cd3b4bd665a.png">

I have try to reproduce with kitty --replay-commands kittyDebug3.txt but I can't see the block cursors again

**Comment by kovidgoyal** [comment]
I'm going to need some way to reproduce to help you.

**Comment by ggsalas** [comment]
@kovidgoyal I've found a way to reproduce:

Open kitty, create 2 windows, change layout, create 1 window more, change other layout -> repeat

![image](https://user-images.githubusercontent.com/8403993/84005990-4b1e6000-a96e-11ea-809a-79faaaf53925.png)
[kittydebug5.txt](https://github.com/kovidgoyal/kitty/files/4744612/kittydebug5.txt)

**Comment by kovidgoyal** [comment]
Can you be a little more detailed please. Start with a kitty command line like:

kitty --config NONE -o enabled_layouts=grid,tall,stack (or wahtever)

then post a set of exact steps to replicate the issue.

**Comment by ggsalas** [comment]
I have run this command: `kitty --config NONE -o enabled_layouts=Fat:bias=85,Tall:bias=65,Stack,Grid,Horizontal,Vertical`

Then press kitty_mod+Enter then kitty_mod+l many times until I get this:
![image](https://user-images.githubusercontent.com/8403993/84008432-2b893680-a972-11ea-9102-838ae5b3907f.png)

**Comment by ggsalas** [comment]
I have tried with `kitty --config NONE -o shell=cat -o enabled_layouts=Fat:bias=85,Tall:bias=65,Stack,Grid,Horizontal,Vertical` and I can't reproduce the bug. 
I'm not sure what it means. I have removed my custom prompt from `.zshrc` file and the bug continue happening.

**Comment by kovidgoyal** [comment]
I cannot reproduce it with zsh, and from your screen shot it looks like
the typical prompt redraw on window resize issue zsh has, see #2712

**Comment by ggsalas** [comment]
Can I see your zsh config? I have started zsh with a `.zshrc` with only 2 lines: ```export PATH="/usr/local/bin:$PATH"
export EDITOR='nvim'
```
and the bug is visible

I'm not sure, but this not seems to be the same bug you mention because is on top of the window, no matter the prompt.
![image](https://user-images.githubusercontent.com/8403993/84059491-010f9b80-a9bb-11ea-8b80-2eaedf8eddf1.png)

**Comment by kovidgoyal** [comment]
It doesnt happen for me with even no .zshrc at all

**Comment by ggsalas** [comment]
I'm not sure to understand. There is no way to fix this? Sorry but I'm confused.

![Screen Recording 2020-06-08 at 19 32 18 2020-06-08 19_35_27](https://user-images.githubusercontent.com/8403993/84062119-459d3600-a9bf-11ea-9c9a-54b024d1d69e.gif)

**Comment by kovidgoyal** [comment]
If it doesn't happen with cat, then it has to be something in your zshrc causing it. I cannto reproduce it even with an empty zshrc and I cannot fix what I cannot reproduce.

**Comment by ggsalas** [comment]
There is a way to clean the window if happens?
The only way I found to disappear is calling the resize command. But after that the bug remains. You can see it on this capture:

![Screen Recording 2020-06-08 at 19 48 55 2020-06-08 19_50_22](https://user-images.githubusercontent.com/8403993/84063556-8138ff80-a9c1-11ea-824d-472b2ee7c552.gif)

**Comment by ggsalas** [comment]
Finaly I found a way to see the bug with cat:
![image](https://user-images.githubusercontent.com/8403993/84064064-54d1b300-a9c2-11ea-9445-491b7e426c19.png)

I have tried with 
```
kitty --config NONE -o shell=cat -o enabled_layouts=Fat:bias=85,Tall:bias=65,Stack,Grid,Horizontal,Vertical
```

![image](https://user-images.githubusercontent.com/8403993/84064194-877bab80-a9c2-11ea-8eaf-3712ad4112f3.png)

1. white `lllllllllllllllllllllllllll` and press enter
2. open new window and do (1) again
3. change layout and repeat (2) 
4. do step 3 10 or 15 times

**Comment by kovidgoyal** [comment]
Nope still does not reproduce for me.

**Comment by ggsalas** [comment]
I hope this video helps: https://drive.google.com/file/d/1AaOKGoiRI2o57NfXRCRxSIfxNGdOEyjH/view?usp=sharing

**Comment by kovidgoyal** [comment]
Sorry I spent half an hour trying and I simply cannot reproduce, and I
cant think of any code path that could result in block cursors being
scattered over the screen.

**Comment by ggsalas** [comment]
Hi @kovidgoyal, I have found some things:
1. this bug disappears with click and drag on the kitty window. You can see this behaviour on this video: https://drive.google.com/file/d/1cgp8QP-3Qqcpm9p3lSKyoH0xD0TtxtB8/view?usp=sharing
I have do a click and drag on vim first to check that only works on an empty kitty window.

2. with `shell=cat` the amount of cursors that appears over the screen is the same amount that the amount of windows opened - 1:
![image](https://user-images.githubusercontent.com/8403993/84510880-11b26100-ac9c-11ea-84d7-3c0d32b7b6d7.png)

3. With macOS 10.15.4 on a macbook pro 15" I see this bug, but with macOS 10.13.6 on a macbook air I can't reproduce the bug. 

I hope this helps. Regards.

**Comment by kovidgoyal** [comment]
If a click and drag fixes it, then that implies those blocks are actually selection blocks, you can confirm that by changing the selection color in kitty.conf with selection_background

**Comment by ggsalas** [comment]
@kovidgoyal Yes, I have changed to green `selection_background #008000` and I get this: 
<img width="1513" alt="Screen Shot 2020-06-13 at 08 09 23" src="https://user-images.githubusercontent.com/8403993/84567183-51855100-ad4d-11ea-899c-871b47ada6cf.png">

**Comment by kovidgoyal** [comment]
Interesting, this definitely looks like a driver bug, basically the buffer used to store which cells are selected is somehow getting ramndomized. Since I cant reproduce its hard for me to investigate, but if you are willing to do some hacking, I can guide you.

**Comment by ggsalas** [comment]
Sure, tell me if I can help.

**Comment by Luflosi** [comment]
I have seen a few one-wide selections on my machine too and I reproduced it with the instructions above:
![screenshot](https://user-images.githubusercontent.com/15217907/84567352-65927800-ad78-11ea-98fd-18f116ef1da5.png)

**Comment by kovidgoyal** [comment]
In line 335 in shaders.c prefix the if condition with true || to make it
always true and see if that fixes it.

**Comment by pyrho** [comment]
I have been experiencing this issue too, but did not open an issue since I have no reliable way of reproducing it.
It ususally happens to me when:
- resizing
- switching layouts

The (pretty shitty) workaround I have is to select some text with the mouse which makes the artifacts go away.

Not certain when this regression appeared but I'm tempted to say around the 0.17.0 release.

**Comment by kovidgoyal** [comment]
Try the code change I posted above.

On Sat, Jun 13, 2020 at 04:19:14AM -0700, Luflosi wrote:
> I have seen a few one-wide selections on my machine too and I reproduced it with the instructions above:
> ![screenshot](https://user-images.githubusercontent.com/15217907/84567352-65927800-ad78-11ea-98fd-18f116ef1da5.png)
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/issues/2747#issuecomment-643609420

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Comment by dmerejkowsky** [comment]
I can confirm that 7824739 fixes the bug here. Thanks!

**Comment by ggsalas** [comment]
I'm trying to test the fix, but when I do make I get this error: ```kitty/data-types.h:11:10: fatal error: 'Python.h' file not found``` I'm sure is my machine, but I can't find a way to compile kitty. Sorry.

**Comment by pyrho** [comment]
@ggsalas please refer to https://sw.kovidgoyal.net/kitty/build.html?highlight=build .
It looks like you are missing `libpython-dev`.

I confirm that I have not been able to reproduce the issue neither after using https://github.com/kovidgoyal/kitty/commit/78247390348444c8c52abc089f3f5dc5f1bcfde7 .

Thanks !

**Comment by ggsalas** [comment]
@pyrho I think I'm missing some python configuration on my machine. This is the error I have when I try to do `make` 
![image](https://user-images.githubusercontent.com/8403993/84595128-c1b6d400-ae2c-11ea-9b95-2579179f1461.png

**Comment by pyrho** [comment]
@ggsalas Installing python3 with [Homebrew](https://brew.sh/) should also install the development headers.
Mine is located here: `/usr/local/Cellar/python/3.7.7`

```
❯ brew info python3
python: stable 3.7.7 (bottled), HEAD
Interpreted, interactive, object-oriented programming language
https://www.python.org/
/usr/local/Cellar/python/3.7.7 (4,090 files, 62.4MB) *
  Poured from bottle on 2020-03-22 at 15:41:06
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/python.rb
==> Dependencies
Build: pkg-config ✘
Required: gdbm ✔, openssl@1.1 ✔, readline ✔, sqlite ✔, xz ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
Python has been installed as
  /usr/local/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /usr/local/opt/python/libexec/bin

You can install Python packages with
  pip3 install <package>
They will install into the site-package directory
  /usr/local/lib/python3.7/site-packages

See: https://docs.brew.sh/Homebrew-and-Python
==> Analytics
install: 184,609 (30 days), 904,076 (90 days), 5,189,932 (365 days)
install-on-request: 153,292 (30 days), 595,046 (90 days), 2,747,963 (365 days)
build-error: 0 (30 days)
                                                                                         
~
❯ cd /usr/local/Cellar/python/3.7.7 && fd Python.h
Frameworks/Python.framework/Versions/3.7/include/python3.7m/Python.h
```


If you are still unable to build, I guess you should open another issue on that topic.

**Comment by rrthomas** [comment]
Brilliant, I was in the middle of reporting this bug, and running master to get a screenshot, and found this fix!

**Comment by flobe** [comment]
7824739 also fixes the bug for me here. Thanks you very much!

**Comment by MichaelAquilina** [comment]
Just came here to report this bug. Experiencing the same issue on 0.17.4. I believe the fix being referenced in this issue has not seen an official release yet? I'll try building on master to see if this fixes the issue in the meanwhile

**Comment by MichaelAquilina** [comment]
Update: can confirm the bug is fixed on master with 7824739. I'll try get this fix packaged in opensuse when a new kitty version is released.
